valid-jsdoc misses issues in function expression docs

With ESLint @ 86d64ca571faa7260233d41b8eb0391a51d59ba0, it looks like the `valid-jsdoc` rule misses issues in documentation for function expressions.

The `script.js` below provides two invalid test cases (both the function declaration and the function expression have a typo in the `@param` name).

```js
/* eslint valid-jsdoc: [2, {requireReturn: false}] */

// valid-jsdoc catches the typo below

/**
 * Check the param name.
 * @param {string} typoInTheName The param description.
 * @return {string} The return description.
 */
function foo(bar) {
  return bar;
}

// valid-jsdoc does not catch the typo below

check(
  /**
   * Check the param name.
   * @param {string} typoInTheName The param description.
   * @return {string} The return description.
   */
  function(bar) {
    return bar;
  });
```

Here's what I get running this through ESLint:
```
$ node bin/eslint.js --no-eslintrc script.js 

/Users/tschaub/projects/eslint/script.js
  5:1  error  Expected JSDoc for 'bar' but found 'typoInTheName'  valid-jsdoc

✖ 1 problem (1 error, 0 warnings)
```

Instead of 1 problem, I would expect two.  The `valid-jsdoc` rule correctly catches the typo in the `@param` name for the function declaration, but misses the same for the function expression.


<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/29912573-valid-jsdoc-misses-issues-in-function-expression-docs?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github).
</bountysource-plugin>

--- a/lib/util/source-code.js
+++ b/lib/util/source-code.js
@@ -238,6 +238,8 @@ SourceCode.prototype = {
                     }
 
                     return parent && (parent.type !== "FunctionDeclaration") ? findJSDocComment(parent.leadingComments, parent.loc.start.line) : null;
+                } else if (node.leadingComments) {
+                    return findJSDocComment(node.leadingComments, node.loc.start.line);
                 }
 
             // falls through

--- a/tests/lib/rules/valid-jsdoc.js
+++ b/tests/lib/rules/valid-jsdoc.js
@@ -41,6 +41,20 @@ ruleTester.run("valid-jsdoc", rule, {
         "/**\n* Description\n* @inheritDoc */\nfunction foo(arg1, arg2){ return ''; }",
         {
             code:
+                "call(\n" +
+                "  /**\n" +
+                "   * Doc for a function expression in a call expression.\n" +
+                "   * @param {string} argName This is the param description.\n" +
+                "   * @return {string} This is the return description.\n" +
+                "   */\n" +
+                "  function(argName) {\n" +
+                "    return 'the return';\n" +
+                "  }\n" +
+                ");\n",
+            options: [{requireReturn: false}]
+        },
+        {
+            code:
                 "/**\n" +
                 "* Create a new thing.\n" +
                 "*/\n" +
@@ -329,6 +343,24 @@ ruleTester.run("valid-jsdoc", rule, {
     ],
 
     invalid: [
+        {
+            code:
+                "call(\n" +
+                "  /**\n" +
+                "   * Doc for a function expression in a call expression.\n" +
+                "   * @param {string} bogusName This is the param description.\n" +
+                "   * @return {string} This is the return description.\n" +
+                "   */\n" +
+                "  function(argName) {\n" +
+                "    return 'the return';\n" +
+                "  }\n" +
+                ");\n",
+            options: [{requireReturn: false}],
+            errors: [{
+                message: "Expected JSDoc for 'argName' but found 'bogusName'.",
+                type: "Block"
+            }]
+        },
         {
             code: "/** @@foo */\nfunction foo(){}",
             errors: [{

--- a/tests/lib/util/source-code.js
+++ b/tests/lib/util/source-code.js
@@ -207,6 +207,36 @@ describe("SourceCode", function() {
 
         });
 
+        it("should get JSDoc comment for FunctionExpression in a CallExpression", function() {
+            var code = [
+                "call(",
+                "  /** Documentation. */",
+                "  function(argName) {",
+                "    return 'the return';",
+                "  }",
+                ");"
+            ].join("\n");
+
+            /**
+             * Check jsdoc presence
+             * @param {ASTNode} node not to check
+             * @returns {void}
+             * @private
+             */
+            function assertJSDoc(node) {
+                var sourceCode = eslint.getSourceCode();
+                var jsdoc = sourceCode.getJSDocComment(node);
+                assert.equal(jsdoc.type, "Block");
+                assert.equal(jsdoc.value, "* Documentation. ");
+            }
+
+            var spy = sandbox.spy(assertJSDoc);
+
+            eslint.on("FunctionExpression", spy);
+            eslint.verify(code, {rules: {}}, filename, true);
+            assert.isTrue(spy.calledOnce, "Event handler should be called.");
+        });
+
         it("should get JSDoc comment for node when the node is a FunctionDeclaration", function() {
 
             var code = [

